src/vm/castcache.h

```diff
@@ -0,0 +1,392 @@
+    }
+
+    FORCEINLINE static TypeHandle::CastResult TryGetFromCacheAny(TypeHandle source, TypeHandle target)
+    {
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+            MODE_ANY;
+        }
+        CONTRACTL_END;
+
+        GCX_COOP();
+
+        return TryGetFromCache(source, target);
+    }
+
+    FORCEINLINE static TypeHandle::CastResult TryGetFromCache(MethodTable* pSourceMT, TypeHandle target)
+    {
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+            MODE_COOPERATIVE;
+        }
+        CONTRACTL_END;
+
+        return TryGetFromCache((TADDR)pSourceMT, target.AsTAddr());
+    }
+
+    static void FlushCurrentCache();
+
+    FORCEINLINE CastCacheEntry* Elements()
+    {
+
+        // element 0 is used for embedded aux data, skip it
+        return (CastCacheEntry*)AuxData() + 1;
+    }
+
+    // TableMask is "size - 1" 
+    // we need that more often that we need size
+    FORCEINLINE DWORD& TableMask()
+    {
+        return *(DWORD*)AuxData();
+    }
+
+    FORCEINLINE BYTE& HashShift()
+    {
+        return *((BYTE*)AuxData() + sizeof(DWORD));
+    }
+
+    FORCEINLINE BYTE& VictimCounter()
+    {
+        return *((BYTE*)AuxData() + sizeof(DWORD) + 1);
+    }
+
+    FORCEINLINE DWORD TableSize()
+    {
+        return this->TableMask() + 1;
+    }
+
+private:
+
+// The cache size is driven by demand and generally is fairly small. (casts are repetitive)
+// Even conversion-churning tests such as Linq.Expressions will not need > 4096
+// When we reach the limit, the new entries start replacing the old ones somewhat randomly.
+// Considering that typically the cache size is small and that hit rates are high with good locality, 
+// just keeping the cache around seems a simple and viable strategy.
+// 
+// Additional behaviors that could be considered, if there are scenarios that could be improved:
+//     - flush the cache based on some heuristics
+//     - shrink the cache based on some heuristics
+// 
+#if DEBUG
+    static const DWORD INITIAL_CACHE_SIZE = 8;    // MUST BE A POWER OF TWO
+    static const DWORD MAXIMUM_CACHE_SIZE = 512;  // make this lower than release to make it easier to reach this in tests.
+#else
+    static const DWORD INITIAL_CACHE_SIZE = 128;  // MUST BE A POWER OF TWO
+    static const DWORD MAXIMUM_CACHE_SIZE = 4096; // 4096 * sizeof(CastCacheEntry) is 98304 bytes on 64bit. We will rarely need this much though.
+#endif
+
+// Lower bucket size will cause the table to resize earlier
+// Higher bucket size will increase upper bound cost of Get
+//
+// In a cold scenario and 64byte cache line:
+//    1 cache miss for 1 probe, 
+//    2 sequential misses for 3 probes, 
+//    then a miss can be assumed for every additional probe.
+// We pick 8 as the probe limit (hoping for 4 probes on average), but the number can be refined further.
+    static const DWORD BUCKET_SIZE = 8;
+
```

Comment:
What evidence do you have for a bucket size of 8?

Reply:
Mostly from guesstimating the number of cache lines that I need to touch in `Get`. I also tried few other numbers. 
Smaller numbers generally have similar performance, but table size would get larger. Bigger numbers eventually make lookup slower. I remember that 16 was definitely slower.

That was before I made some improvements, so it may make sense to try again. 

There might not be a lot to gain though. While filling up the table, collisions are rare and buckets are small. Once we start getting collisions,it means the table is getting full, buckets start growing quickly and we resize. So +/- 1 in the threshold value does not change a lot unless we are "lucky" to stay exactly at threshold. 
Where it does matter is when we are at max cache size, then it really correlates with the average number of reprobes. 

I will run with few more bucket sizes. I am curious too.

---
In reply to: [306591847](https://github.com/dotnet/coreclr/pull/23548#discussion_r306591847) [](ancestors = 306591847)